src/logger.ts

```diff
@@ -53,16 +56,19 @@ export class Logger {
   static DEFAULT_OPTIONS:
       Readonly<LoggerConfig> = {level: 'error', levels: LEVELS, tag: ''};
 
-  // TODO: Mark this private when TypeScript 2.9 comes out.
-  // See https://github.com/Microsoft/TypeScript/issues/20080 for more
-  // information.
-  [kTag]: string;
+  private[kTag]: string;
 
-  /**
-   * Emits a log at this log level.
-   */
-  // tslint:disable-next-line:no-any
-  [logLevel: string]: (...args: any[]) => this;
+  // ts: The compiler can't statically detect that these will be definitely
+  // assigned. They will be if default log levels are used, so we use non-null
+  // annotations here.
+  // tslint:disable:no-any
+  silent!: (...args: any[]) => this;
```

Comment:
I'm still not convinced that we should use these non-null annotations. If an instance of this class is created with custom levels, some (or any) of these won't be set, right? How do we prevent coding errors in that case?

Reply:
The alternative, if we want to externalize log functions, is to also have a second `DefaultLevelsLogger` in addition to `CustomLevelsLogger`. This works but also means that any time a logger is constructed, it needs to be casted to one of the two (otherwise it has no public API). I've been trying to finagle with TypeScript to discover other alternatives (like constructor overloading) but there aren't any AFAIK.

Personally I think we should get rid of custom levels, as they are virtually unused in practice. For client libraries we don't use them -- and I don't see a reason for third-party modules to have a direct dependency on this module for its logger. Because the role of the `nodejs-common` module as a collection of helpers for GCP client libraries, I think that if someone is actually using `Logger` for other reasons, then they are on their own.

Reply:
What about always adding default log levels regardless of whether they are included in the custom options?

Reply:
That seems like a semver-major change. The library is experimental, so it's probably reasonable to do so (ping @JustinBeckwith) -- but I think if we were to make a semver-major change, we might as well remove custom log levels entirely instead.

Reply:
I think though that using the default log levels should still be considered a coding error, regardless of whether they're added by default.

Reply:
I thought the default log levels were supposed to be always there even though you were using custom log levels. I think I thought that way because the non-null operators were put to those default log functions.

In that case, I think we should have a new base class (say `BaseLogger`), from which both the `Logger` class and the `CustomLevelsLogger` class should inherit. `Logger` will have these logging functions by default while `BaseLogger` and `CustomLevelsLogger` won't. WDYT?

BTW if this PR blocks your work, I can LGTM as is.

Reply:
That makes sense. I think I prototyped something similar when I was working on this PR -- I'll try it out again. (Though, to keep things stable, `Logger` will be the one that will stay the same, not `BaseLogger`.)